Switch from "Title Case" to "Sentence case"

marusak opened the pull request:
First step in #14285 

Opening PR mostly to get reviewd the second commit. Have I changed any string that I should not? Are there some changes we don't want or should be different?

There are still some title case strings that were not changed: (I think it is correct they stayed unchanged)
```
"Web Console for Linux servers"
"Web Console"
"To get software updates, this system needs to be registered with Red Hat, either using the Red Hat Customer Portal or a local subscription server."
"Start Docker"
"Remove Tang keyserver"
"Once Cockpit is installed, enable it with \"systemctl enable --now cockpit.socket\"."
"Edit Tang keyserver"
"Authentication is required to perform privileged tasks with the Cockpit Web Console"
"Administration with Cockpit Web Console"
"About Web Console"
```

@garrett

garrett commented on pkg/docker/manifest.json.in:
```diff
@@ -6,7 +6,7 @@
     "menu": {
         "index": {
-            "label": "Docker Containers",
+            "label": "Docker containers",
```
PF4 says "Write menu/navigation items and their corresponding page titles in title case." @ https://www.patternfly.org/v4/design-guidelines/content/capitalization#capitalization-across-products — I'm not sure if I agree with that. Let's see what it looks like.

(We may want to go over this PR with this in mind if we do want to have navigation remain Title Case.)

marusak commented on pkg/docker/org.cockpit-project.cockpit-docker.metainfo.xml:
```diff
@@ -2,7 +2,7 @@
 <component type="addon">
   <id>org.cockpit_project.cockpit_docker</id>
   <metadata_license>CC0-1.0</metadata_license>
-  <name>Docker Containers</name>
```
According to [guidelines](https://www.patternfly.org/v4/design-guidelines/content/capitalization#capitalization-across-products) this should stay titlecase. There is at least one or two more examples  of the same problem. We might need just list of strings that should not be touched.

marusak commented on tools/title2sentance.py:
```diff
@@ -0,0 +1,94 @@
+    'Web Console',
+    'Red Hat',
+    'Docker',
+    'CPU',
+    'IP',
+    'VNC',
+    'VM',
+    'LVM',
+    'STP',
+    'SSH',
```
WE maybe could have a logic that if single word is all caps, just keep it all caps...

garrett commented on pkg/lib/machine-info.js:
```diff
@@ -67,26 +67,26 @@ const chassis_types = [
-    _("Docking Station"),
-    _("All In One"),
-    _("Sub Notebook"),
+    _("Hand held"),
```
This should probably be "Handheld".

marusak commented on tools/title2sentance.py:
```diff
@@ -0,0 +1,94 @@
+    'DNS',
+    'Cockpit',
+    'Linux',
+    ': Server',
+    ': Invalid',
+    '. Its',
```
Strings that are more than one sentence are a bit trickier... This is just simple hack because there is just a few of such cases.

garrett commented on pkg/lib/machine-info.js:
```diff
@@ -67,26 +67,26 @@ const chassis_types = [
-    _("Sub Notebook"),
+    _("Hand held"),
+    _("Docking station"),
+    _("All in one"),
```
Should be "All-in-one".

garrett commented on pkg/lib/machine-info.js:
```diff
@@ -67,26 +67,26 @@ const chassis_types = [
+    _("Hand held"),
+    _("Docking station"),
+    _("All in one"),
+    _("Sub notebook"),
```
Should be "Sub-notebook"

marusak commented on Makefile.am:
```diff
@@ -436,6 +436,12 @@ distcheck-hook::
+title-case: po/cockpit.pot
+	$(MSGGREP) --no-wrap --msgid -e '\b[A-Z][a-z]\+\b \+\b[A-Z][a-z]' po/cockpit.pot | grep "^msgid " | cut -d " " -f2- | sort --reverse > titles
+
+title2sentance: title-case
+	tools/title2sentance.py -i titles -o t2s.sh
```
yes yes, this is not yet finished :)

martinpitt replied:
sent"e"nce.

garrett commented on pkg/lib/machine-info.js:
```diff
@@ -67,26 +67,26 @@ const chassis_types = [
+    _("Lunch box"), /* 0x10 */
+    _("Main server chassis"),
+    _("Expansion chassis"),
+    _("Sub chassis"),
```
"Sub-chassis", unless it's the chassis of a submarine. :wink:

garrett commented on pkg/lib/notifications.js:
```diff
@@ -20,10 +20,10 @@
-example, the "Software Updates" page can send out a notification when
+example, the "Software updates" page can send out a notification when
 it detects that software updates are available.  The shell will then
-highlight the menu entry for "Software Updates" and the "System"
-overview page will also mention it in its "Operating System" section.
+highlight the menu entry for "Software updates" and the "System"
```
If we do adopt the navigation should have Title Case suggestion from PF4, then these should be kept as-is. (It's a comment, so it doesn't really matter much though, but still.)

garrett commented on pkg/machines/README.md:
```diff
@@ -16,7 +16,7 @@ In the vagrant:
         -f ${NAME}.img -n ${NAME}   # to create and start a dummy VM
 
-Log into Cockpit as the 'root' user (pwd 'foobar') and follow `Virtual Machines` from the left-side menu.
+Log into Cockpit as the 'root' user (pwd 'foobar') and follow `Virtual machines` from the left-side menu.
```
Kept as Title Case, if we adopt the navigation as title case PF4 recommendation.

garrett commented on pkg/packagekit/index.html:
```diff
@@ -20,7 +20,7 @@
 
 <head>
-  <title translate>Software Updates</title>
+  <title translate>Software updates</title>
```
If we adopt the navigation as title case PF4 recommendation, then this should remain Title Case.

garrett commented on pkg/packagekit/manifest.json.in:
```diff
@@ -8,7 +8,7 @@
     "tools": {
         "index": {
-            "label": "Software Updates",
+            "label": "Software updates",
```
If we adopt the navigation as title case PF4 recommendation, then this should remain Title Case.

garrett commented on pkg/sosreport/manifest.json.in:
```diff
@@ -6,7 +6,7 @@
     "tools": {
         "index": {
-            "label": "Diagnostic Reports",
+            "label": "Diagnostic reports",
```
If we adopt the navigation as title case PF4 recommendation, then this should remain Title Case.

garrett commented on pkg/sosreport/org.cockpit-project.cockpit-sosreport.metainfo.xml:
```diff
@@ -2,7 +2,7 @@
   <id>org.cockpit_project.cockpit_sosreport</id>
   <metadata_license>CC0-1.0</metadata_license>
-  <name>Diagnostic Reports</name>
+  <name>Diagnostic reports</name>
```
If we adopt the navigation as title case PF4 recommendation, then this should remain Title Case.

garrett reviewed (commented):
This looks fantastic!

I've only looked through the changes in the PR. I'm still waiting for another branch to build, so I'll check this in PR Cockpit later today.

I've added some comments as I've gone through, but they're just notes on things that _might_ need changing (and some of them would probably be in another PR anyway, such as the hyphenation of machine types).

The biggest things that _might_ need changing are navigation-related titles, if we adopt the PF recommendation of having navigation remain Title Case. This would mainly impact the `manifest.json.in` files and possibly content in `<title></title>` in HTML pages. We should probably discuss and think about this though.

marusak commented:
To commpare, screenshot from master vs this PR. (Ignore Image builder and perf. graphs as those are external ones). Also name on tab changes.

Currently:
![uppercase](https://user-images.githubusercontent.com/12330670/92381194-ae08ca00-f10a-11ea-9aa9-0cf3a3bfb732.png)

This PR:
![lowercase](https://user-images.githubusercontent.com/12330670/92381209-b5c86e80-f10a-11ea-9497-3b1d6b568c43.png)

garrett commented:
Hmm. I might like the titlecase for page names in the navigation (and page title). But it's nice to see it lowercased everywhere else (aside from "Check For Updates", which is an oversight).

(At least page names in the navigation are controlled by the schema files, so they're easy to switch back, for the most part.)

Building locally here now. I'll make a list of findings.

garrett commented:
## Log in page

- [ ] "Other Options"

## Storage

- [ ] "NFS Mounts"
- [ ] "Create RAID Device" (in the menu)
- [ ] "Create RAID Device" (dialog title)
- [ ] "RAID Level"
- [ ] "___ GiB Encrypted partition of" ("Encryption" should be lowercase here)

## Networking
- [ ] "VLAN Settings" (dialog title)
- [ ] "VLAN Id" (should be "VLAN ID")

## Virtual Machines
- [ ] "CPU Type" (overview tab on VMs)
- [ ] "____ Memory Adjustment" (dialog title)
- [ ] "____ VCPU details" (should be "vCPU detauls")
- [ ] "vCPU Count"
- [ ] "vCPU Maximum"
- [ ] "MAC Address"
- [ ] "IP Address" (should be "IP address")
- [ ] "___ Virtual Network Interface Settings"
- [ ] "MAC Address"

## Services
- Some service titles are "Title Caps", some are "Sentence caps", some are lowercase... I think these come from the various systemd services.

## Applications

- [ ] "View Project Website"

## Development :wink: 
- [ ] "Development playground"
- [ ] Pretty much all the links here
- [ ] "Page Status" 
- [ ] Most lot of the patterns are still wrong

# SELinux
- [ ] "SELinux Policy"

## Software updates
- I can't see this locally, as I'm on Silverblue and haven't built a VM for this branch yet. :grin: 
- [ ] "Check for Updates" button (according to the screenshots above)

marusak commented:
Thanks @garrett for going through the pages! I will go through all of those manually and check what is the problem and will edit your comment to keep it updated.
For now I see one issue - the regex we use for finding title case is not good enough - it requires "capital letter followed by bunch of lower case latters..." which is not the case for example in `NFS Mounts`, `MAC Address`... as well as the sentence does not need to start with capital letter like in `vCPU Count`. I will play with this and see what would work - maybe making the regex more robust, or less strict or possibly dropping it all together and just applying some logic to all strings?

garrett commented:
@marusak Well, we could make more than one pass. One to catch most of the stuff, an additional one (or multiple additional passes) to clean up the corner cases.

It's probably better to go with a few simple regular expressions than a complicated one that's not so understandable.

But hopefully soon, we'll just merge the results and can forget about the script. I consider it throwaway code, for a purpose, so it's fine by me if it's messy if it has a proper output.

marusak commented:
I just pushed version two. See [diff](https://github.com/cockpit-project/cockpit/compare/c7b4ded6dc1d6d78bfe36e622e872280df01e954..1a8359ab6597646272cd8af04d665a2af3b417e5) for easier review of changes.
(I immediately see there `packagekit` problem, will fix and push right away)

marusak commented:
Ok, [this diff](https://github.com/cockpit-project/cockpit/compare/c7b4ded6dc1d6d78bfe36e622e872280df01e954..906d04607f92161d987b96ff6aead40e1283baf1) is the one you want to review. I had a few more fixups.
Now we need to decide if we want to title case the menu or now.
This also has potential to break test, so spinning one.

marusak commented:
Rebased and fixed some forgotten untranslated strings.
From the list @garrett provided only one thing is remaining:
```
___ GiB Encrypted partition of" ("Encryption" should be lowercase here)
```
I am not sure where this is in the UI. These strings are defined in `pkg/storaged/utils.js` and from that point of view they are correct. If they are always used with some prefix they need to be used differently.

marusak commented:
Oh and forgotten - this PR is ready for review and hopefully close to landing :)

garrett commented:
>> ___ GiB Encrypted partition of" ("Encryption" should be lowercase here)
>
> I am not sure where this is in the UI. These strings are defined in pkg/storaged/utils.js and from that point of view they are correct. If they are always used with some prefix they need to be used differently.

It's in the disk details, under "Content".

![Screenshot_2020-09-09 Storage - garrett Rain](https://user-images.githubusercontent.com/10246/92575674-ec2bf800-f288-11ea-8744-bea6073cf06c.png)

All of the strings here need to be proper cased. It's supposed to be "$AMOUNT $UNIT encrypted..." (where the # serves as the "capital" in that phrase). Also, the same for "Physical Volume"... that should be "physical volume". And "File System" should be "file system" (or "filesystem" — it's debatable, so "file system" with two words is probably preferable for us, but it's a matter of opinionated style).

martinpitt commented on Makefile.am:
```diff
@@ -436,6 +436,12 @@ distcheck-hook::
 	tools/make-bots
 
+title-case: po/cockpit.pot
+	$(MSGGREP) --no-wrap --msgid -e '\S\+\s' po/cockpit.pot | grep "^msgid " | cut -d " " -f2- | sort --reverse > titles
```
This doesn't filter for title cases, just pretty much anything which is not just a single word. A better name would be good to avoid confusion, like "multi-word" or whatever your intention is?

marusak replied:
oh right, it used to do that but then I just changed it to pick all strings where it might make sense to wonder about "title" vs "sentence" case. (As `VLAN Settings` or `vCPU Settings` or even `$0 Settings` were not picked as title case).

marusak replied:
changed to `multi-word`

martinpitt commented on pkg/apps/application.jsx:
```diff
@@ -100,7 +100,7 @@ export class Application extends React.Component {
             if (!comp) {
                 if (metainfo_db.ready)
-                    return <div>{_("Unknown Application")}</div>;
+                    return <div>{_("Unknown application")}</div>;
```
So this "Change strings from sentance case to title case" commit was done entirely automatically with the script? Or does it still have some manual adjustments? The commit message should point out exactly how it was done, so that it can be reproduced.

Also, it would be really good to adjust the msgids in *.po as well in the same commit.

marusak replied:
> So this "Change strings from sentance case to title case" commit was done entirely automatically with the script?

yes. Well I did `make title2senance` and then just ran `t2s.sh` and commit what it changed. But no manual intervention.

> Also, it would be really good to adjust the msgids in *.po as well in the same commit.

I guess I could do that (with un-wrapping it first). I thought I would do it just in the weblate repo and then sync it back here. But if this is acked I can just run that `t2s.sh` script on `po/` no problem with that.

martinpitt reviewed (changes requested):
Nice work, thank you! Feel free to send a separate PR for the string fixes, so that what remains is just the automatic part. I didn't review the python script thoroughly, as I assume we'll remove it again once the transition is complete.

But adjusting the *.po files is a blocker -- we can't break *all* translations unnecessarily in one go.

marusak commented:
@martinpitt I went rouge and skipped all distropkg tests on non-appstream tests. WDYT?

martinpitt commented on test/verify/check-accounts:
```diff
@@ -24,6 +24,7 @@ from testlib import *
 
 
 @nondestructive
+@skipImage("Do not test cockpit package", "rhel-8-3-distropkg")
```
How about "Do not test BaseOS packages"? "cockpit package" is a bit confusing I think, as everything here is about cockpit.

martinpitt reviewed (commented):
I'm fine with these skips. Just one nitpick, otherwise this LGTM! Thanks!

martinpitt reviewed (dismissed):
Dakujem!

martinpitt commented:
[testSnapshotRevert failure](https://logs-https-cockpit.apps.ci.centos.org/logs/pull-14574-20200910-074620-24ec65df-debian-testing/log.html#200-2), same on [fedora-33](https://logs.cockpit-project.org/logs/pull-14574-20200910-074620-24ec65df-fedora-33/log.html#200-2). This is likely real (snapshots only work on these two OSes)

garrett commented:
To cover PF4's guidelines of navigation elements being in Title Case (one of the very few exceptions), I did `git co master **/manifest.json.in`

However, this doesn't solve the `<title>` case nor does it address breadcrumbs. :shrug: 

Otherwise, everything looks good.

garrett commented:
Looking at things, I think the big exception is the VM page, but we can consider that as "virtual machines" just being capitalized, I suppose. Everything else has Title Case == Sentence case, as it's one word.

garrett commented:
BTW: Here's that link again on capitalization: https://www.patternfly.org/v4/design-guidelines/content/capitalization

The quote:
> -    **Write menu/navigation items and their corresponding page titles in title case**.
> -    Write products, services, and features in title case.
> -    Write all other areas (like button text, form field labels, and empty state headings) in sentence case. 

And:

> Usually, page titles appear in a breadcrumb trail. If all page titles are in title case, they should appear in title case in the breadcrumb trail. If all page titles are in sentence case, they should also appear in sentence case in the breadcrumb trail. 

(Which I didn't fix. Perhaps we should fix that. It's debatable here, I think, as we're talking about virtual machines, not necessarily the Virtual Machines page.)

garrett commented:
I still have "Encrypted data" and "Physical volume":

![Screenshot_2020-09-10 Storage - garrett Rain](https://user-images.githubusercontent.com/10246/92713773-ac821080-f35b-11ea-81c3-237ad2b7eb6d.png)

(It's in the disk details list on the Storage page.)

marusak commented on pkg/kdump/manifest.json.in:
```diff
@@ -6,7 +6,7 @@
 
     "tools": {
         "index": {
-            "label": "Kernel dump",
```
Thanks @garrett for the commit, but I'll have to drop it and do it properly. We need to also change it in .po files and also in weblate repo, so I want script to do it and not do it manually.
I'll do it now.

garrett reviewed (approved):
> will be follow up for sure

Sure. This PR is close enough to merge, IMO.

Cockpit has had a bunch of title/sentence case issues for a while, so this is a major improvement, even if there are a few stragglers.

I'm happy with what few little ones remain to be handled in follow-up PRs at this point.
